### PR TITLE
BaseModel class

### PR DESCRIPTION
**Pull request by PososikTeam**

### CODE REVIEW

File: src/models/base.py [hunk 1]
```diff
@@ -213,22 +17,4 @@ def get_forward_output_channels(self) -> Union[int, List[int]]:
...
-        return forward_channels, hooks_channels
-
-    @property
-    def features_info(self):
-        return self.__features_info
-
-    @property
-    def feature_hooks(self):
-        return self.__feature_hooks
+
```
Comment (VladVin) [inline src/models/base.py:20]: Add abstract method `no_weight_decay`, look for the signature in `Constructor`. And remove the `hasattr` check there

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/backbones/base_backbone.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    @abstractmethod
+    def forward_features(self, *args, **kwargs):
+        """Forward method for getting backbone feature maps.
+           They are mainly used for segmentation and detection tasks.
+        """
+        pass
+
+    @abstractmethod
+    def get_forward_feature_output_channels(self) -> Union[int, List[int]]:
```
Comment (VladVin) [inline src/models/backbones/base_backbone.py:23]: Maybe give a shorter name `get_forward_features_channels`?

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/tasks/base.py [hunk 3]
```diff
@@ -41,9 +41,17 @@ def forward_with_gt(batch: Dict[str, Any]) -> Dict[str, torch.Tensor]:
         """Abstract forward method for training(with ground truth labels)."""
         pass
 
+    @abstractmethod
+    def no_weight_decay(self) -> List[str]:
```
Comment (VladVin) [inline src/tasks/base.py:45]: It should be not here but in `BaseModel`. Otherwise, a user has to setup no-weight-decay params in the task, but they are defined for each model custom which is overkill

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/backbones/base_backbone.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from abc import ABC, abstractmethod
+from typing import List, Union, Optional
+
+from src.models.base import BaseModel
+
+
+class BaseBackbone(BaseModel, ABC):
+    """Base model for TorchOk Backbones"""
+    def __init__(self,
+                 in_channels: Optional[Union[int, List[int]]] = None,
```
Comment (VladVin) [inline src/models/backbones/base_backbone.py:10]: You can add `Tuple[int, ...]` into `Union` because it's also acceptable

File: src/models/backbones/base_backbone.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    @abstractmethod
+    def forward_features(self, *args, **kwargs):
+        """Forward method for getting backbone feature maps.
+           They are mainly used for segmentation and detection tasks.
+        """
+        pass
+
+    @property
+    def out_feature_channels(self):
```
Comment (VladVin) [inline src/models/backbones/base_backbone.py:31]: No typings

File: src/models/backbones/resnet.py [hunk 6]
```diff
@@ -143,17 +139,25 @@ def forward(self, x: torch.Tensor):
...
-        return features_info
-
-    def get_forward_output_channels(self) -> Union[int, List[int]]:
-        """Return number of output channels."""
-        return self.num_features
+    def forward_features(self, x: torch.Tensor):
+        """Forward method."""
+        features = [x]
+        x = self.convbnact(x)
+        x = self.maxpool(x)
```
Comment (VladVin) [inline src/models/backbones/resnet.py:146]: Shouldn't we also return the feature map after the "stem" layer?

File: src/models/backbones/resnet.py [hunk 7]
```diff
@@ -143,17 +139,25 @@ def forward(self, x: torch.Tensor):
...
-        features_info = []
-        for num_channels, module_name, stride in zip(output_channels, module_names, strides):
-            feature_info = FeatureInfo(module_name=module_name, num_channels=num_channels, stride=stride)
-            features_info.append(feature_info)
-        return features_info
-
-    def get_forward_output_channels(self) -> Union[int, List[int]]:
-        """Return number of output channels."""
-        return self.num_features
+    def forward_features(self, x: torch.Tensor):
```
Comment (VladVin) [inline src/models/backbones/resnet.py:142]: No typings

File: src/models/base.py [hunk 8]
```diff
@@ -1,234 +1,46 @@
...
-    
-    def get_output_channels(self) -> Tuple[Union[int, List[int]], List[int]]:
-        """Create forward an hooks channels numbers.
-        
-        Returns:
-            forward_channels: Forward pass output channels.
-            hooks_channels: Hooks output channels.
+
+    def no_weight_decay(self) -> List[str]:
+        """Create module names for which weights decay will not be used.
```
Comment (VladVin) [inline src/models/base.py:28]:
```suggestion
        """Create module names for which weight decay will not be used.
```

File: src/models/identity.py [hunk 9]
```diff
@@ -1,17 +0,0 @@
-from torch import nn, Tensor
-
-from src.models.base import BaseModel
-
-
-class Identity(BaseModel):
```
Comment (VladVin) [inline src/models/identity.py:6]: Why did you remove an Identity module? It should be in the backbones, heads and pooling registries

File: src/tasks/base.py [hunk 10]
```diff
@@ -42,9 +42,19 @@ def forward_with_gt(batch: Dict[str, Any]) -> Dict[str, torch.Tensor]:
         """Abstract forward method for training(with ground truth labels)."""
         pass
 
+    def train_modules(self) -> List[str]:
+        """Create a list of modules that need to train. Method used to apply optimizer for returned modules.
```
Comment (VladVin) [inline src/tasks/base.py:46]:
```suggestion
        """Create a list of 1st-level modules that need to be optimized. The method is used to apply an optimizer for the returned modules.
```

File: src/tasks/base.py [hunk 11]
```diff
@@ -42,9 +42,19 @@ def forward_with_gt(batch: Dict[str, Any]) -> Dict[str, torch.Tensor]:
...
+
+        By default, it would be self.children().
+        
+        Returns: List of modules that need to train.
```
Comment (VladVin) [inline src/tasks/base.py:50]:
```suggestion
        Returns: List of modules that need to be optimized.
```

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/identity.py (hunk 9 above)
Comment (PososikTeam) [reply src/models/identity.py:6]: Because we decide not to use Identity in tasks, fro example in classification task we use
`neck_name = self._hparams.task.params.get('neck_name')` for get neck name but we can use
`neck_name = self._hparams.task.params.get('neck_name', 'Identity')`. So why we need this modules?

File: src/models/backbones/resnet.py (hunk 6 above)
Comment (PososikTeam) [reply src/models/backbones/resnet.py:146]: In previous version of TorchOk it not returned, but i think we should to give flexibility

File: src/models/base.py [hunk 12]
```diff
@@ -1,234 +1,50 @@
...
         """
-        forward_channels = self.get_forward_output_channels()
-        hooks_channels = self._get_hooks_output_channels()
-        return forward_channels, hooks_channels
+        return list()
 
     @property
-    def features_info(self):
-        return self.__features_info
+    def in_channels(self) -> Union[int, List[int]]:
```
Comment (VladVin) [inline src/models/base.py:35]: `Optional[Union[int, List[int], Tuple[int, ...]]]`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: environment.yml [hunk 13]
```diff
@@ -16,18 +15,6 @@ dependencies:
   - cudatoolkit=11.1
   - pytorch-lightning=1.6.2
   - torchmetrics=0.9.2
-=======
-  - mlflow=1.22.0
-  - pandas=1.4.3
-  - pillow=9.1.1
-  - pip=22.1.2
-  - python=3.9
-  - pytorch=1.12
```
Comment (VladVin) [inline environment.yml:25]: Merge dependencies correctly

**Review by VladVin — CHANGES_REQUESTED** [review]
